File: backend/app/schemas/ai.py
"""Pydantic models for AI enrichment and suggestion endpoints."""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.constants import validate_item_type


# ── Request models ─────────────────────────────────────────────────────────────


class EnrichRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=200, description="Name of the item")
    destination: str = Field(..., min_length=1, max_length=120, description="City or country")
    item_type: str = Field(..., description="One of: attraction, restaurant, hotel, transport, activity")

    @field_validator("item_type")
    @classmethod
    def check_item_type(cls, v: str) -> str:
        return validate_item_type(v)


class EnrichBatchRequest(BaseModel):
    items: list[EnrichRequest] = Field(..., min_length=1, max_length=5)


class SuggestionsRequest(BaseModel):
    plan_id: str = Field(..., description="UUID of the plan")
    force_refresh: bool = Field(False, description="Skip cache and generate fresh suggestions")
    exclude_names: list[str] = Field(default_factory=list, description="Names already shown or added")


class NextSuggestionRequest(BaseModel):
    plan_id: str = Field(..., description="UUID of the plan")
    exclude_names: list[str] = Field(default_factory=list, description="Names already shown or added")
    exclude_slugs: list[str] = Field(
        default_factory=list,
        description="Canonical slugs already shown — guards against alias collisions",
    )


class CrossCityTransportRequest(BaseModel):
    plan_id: str = Field(..., description="UUID of the plan")


TransportMode = Literal["drive", "train", "bus", "ferry", "flight"]


class TransportOption(BaseModel):
    """A single cross-city transport option resolved from a real routing source.

    `is_estimate=True` means the duration came from the haversine + cruise-speed
    estimator (flight only); the UI shows an explicit "estimate" chip in that case.
    `transit_summary` is a short carrier label like "ICE 100 + Bus 232" that
    Transitous attaches to train/bus/ferry results.
    """
    mode: TransportMode
    name: str
    duration_seconds: int | None = None
    distance_meters: int | None = None
    is_estimate: bool = False
    transit_summary: str | None = None
    geometry: list[list[float]] | None = None


class TransportSuggestionItem(BaseModel):
    source_item_id: str | None = None
    source_item_title: str | None = None
    source_item_location: str | None = None
    source_destination_id: str | None = None
    destination_item_id: str | None = None
    destination_item_title: str | None = None
    destination_item_location: str | None = None
    destination_destination_id: str | None = None
    scope: str | None = None
    source_day_number: int | None = None
    destination_day_number: int | None = None
    source_city: str | None = None
    destination_city: str | None = None
    source_country: str | None = None
    destination_country: str | None = None
    options: list[TransportOption] = Field(default_factory=list)


class TransportSuggestionsResponse(BaseModel):
    suggestions: list[TransportSuggestionItem]


# ── Response models ────────────────────────────────────────────────────────────


class EnrichedItemResponse(BaseModel):
    # Stable fields (always present after enrichment)
    place_id: str | None = None
    canonical_name: str | None = None
    description: str | None = None
    location: str | None = None
    image_url: str | None = None
    lat: float | None = None
    lng: float | None = None
    timezone: str | None = None
    categories: list[str] | None = None

    # Per-type fresh fields — all optional so one model covers all five types
    opening_hours: str | None = None
    price_range: str | None = None
    tips: list[str] | None = None
    cuisine: str | None = None
    reservation_tips: str | None = None
    amenities: list[str] | None = None
    check_in_time: str | None = None
    booking_tips: str | None = None
    schedule: str | None = None
    duration: str | None = None


class AiSuggestionItemResponse(BaseModel):
    name: str
    item_type: str
    one_line: str | None = None
    price_hint: str | None = None
    cached: bool = False
    slug: str | None = None
    destination_city: str | None = None
    enriched: EnrichedItemResponse | None = None


class AiSuggestionsResponse(BaseModel):
    suggestions: list[AiSuggestionItemResponse]

File: frontend/src/lib/api/ai.ts
import { apiFetch, apiSse } from "./client";

export interface EnrichedItem {
  place_id: string | null;
  description: string | null;
  opening_hours: string | null;
  price_range: string | null;
  tips: string[] | null;
  cuisine: string | null;
  reservation_tips: string | null;
  amenities: string[] | null;
  check_in_time: string | null;
  booking_tips: string | null;
  schedule: string | null;
  duration: string | null;
  location: string | null;
  image_url: string | null;
  lat: number | null;
  lng: number | null;
  timezone: string | null;
  categories: string[] | null;
}

export interface CrossCityMarker {
  cross_city_pair: string;
}

export type SameDayTransportMode = "walk" | "bike" | "drive" | "transit";

export type CrossCityTransportMode = "drive" | "train" | "bus" | "ferry" | "flight";

export interface CrossCityTransportData {
  cross_city_pair: string;
  source_destination_id: string;
  destination_destination_id: string;
  mode: CrossCityTransportMode;
  duration_seconds: number | null;
  distance_meters: number | null;
  is_estimate: boolean;
  transit_summary: string | null;
}

export interface SameDayTransportData {
  same_day_pair: string;
  mode: SameDayTransportMode;
  distance_meters: number;
  duration_seconds: number;
  transit_summary?: string;
  geometry?: [number, number][];
}

export interface PlaceSuggestion {
  slug: string;
  name: string;
  destination: string;
  item_type: string;
  description: string | null;
  location: string | null;
  image_url: string | null;
  lat: number | null;
  lng: number | null;
}

export interface AiSuggestion {
  name: string;
  item_type: string;
  one_line: string | null;
  price_hint: string | null;
  slug: string;
  cached: boolean;
  destination_city: string | null;
  enriched?: EnrichedItem | null;
}

export interface AiSuggestionsResult {
  suggestions: AiSuggestion[];
}

export interface TransportOption {
  mode: CrossCityTransportMode;
  name: string;
  duration_seconds: number | null;
  distance_meters: number | null;
  is_estimate: boolean;
  transit_summary: string | null;
  geometry: [number, number][] | null;
}

export interface TransportSuggestion {
  source_item_id: string | null;
  source_item_title: string | null;
  source_item_location: string | null;
  source_destination_id: string | null;
  destination_item_id: string | null;
  destination_item_title: string | null;
  destination_item_location: string | null;
  destination_destination_id: string | null;
  scope: "cross_city" | null;
  source_day_number: number | null;
  destination_day_number: number | null;
  source_city: string | null;
  destination_city: string | null;
  source_country: string | null;
  destination_country: string | null;
  options: TransportOption[];
}

export interface TransportSuggestionsResult {
  suggestions: TransportSuggestion[];
}

export const enrichItem = (
  name: string,
  destination: string,
  item_type: string,
  signal?: AbortSignal,
): Promise<EnrichedItem> =>
  apiFetch<EnrichedItem>("/ai/enrich", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name, destination, item_type }),
    signal,
  });

export const enrichBatch = (
  items: Array<{ name: string; destination: string; item_type: string }>,
  signal?: AbortSignal,
): Promise<EnrichedItem[]> =>
  apiFetch<EnrichedItem[]>("/ai/enrich-batch", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ items }),
    signal,
  });

export const autocompletePlaces = (
  q: string,
  destination: string,
  item_type: string,
  signal?: AbortSignal,
): Promise<PlaceSuggestion[]> => {
  const params = new URLSearchParams({ q, destination, item_type });
  return apiFetch<PlaceSuggestion[]>(`/places/autocomplete?${params.toString()}`, { signal });
};

export const getSuggestions = (
  planId: string,
  forceRefresh = false,
  excludeNames?: string[],
  signal?: AbortSignal,
): Promise<AiSuggestionsResult> =>
  apiFetch<AiSuggestionsResult>("/ai/suggestions", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      plan_id: planId,
      force_refresh: forceRefresh,
      ...(excludeNames?.length ? { exclude_names: excludeNames } : {}),
    }),
    signal,
  });

export const getNextSuggestion = (
  planId: string,
  excludeNames: string[],
  excludeSlugs: string[] = [],
  signal?: AbortSignal,
): Promise<AiSuggestion> =>
  apiFetch<AiSuggestion>("/ai/suggestions/next", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      plan_id: planId,
      exclude_names: excludeNames,
      exclude_slugs: excludeSlugs,
    }),
    signal,
  });

export const streamCrossCityTransportSuggestions = (
  planId: string,
  onPair: (pair: TransportSuggestion) => void,
  signal?: AbortSignal,
): Promise<void> => {
  const params = new URLSearchParams({ plan_id: planId });
  return apiSse<TransportSuggestion>(
    `/ai/transport-suggestions/stream?${params.toString()}`,
    (name, data) => {
      if (name === "pair") onPair(data);
    },
    signal,
  );
};

